pkg/pillar/cmd/baseosmgr/handlebaseos.go

```diff
@@ -1018,3 +1023,169 @@ func handleOtherPartRebootReason(ctxPtr *baseOsMgrContext, status *types.BaseOsS
+		saveConfigRetryUpdateCounter(ctxPtr)
+		log.Noticef("UpdateRetry from %s to %s",
+			partStatus.ShortVersion, failedPartStatus.ShortVersion)
+		zboot.SetOtherPartitionStateUpdating(log)
+		updateAndPublishZbootStatus(ctxPtr, failedPartStatus.PartitionLabel, false)
+		baseOsStatus := lookupBaseOsStatusByPartLabel(ctxPtr, failedPartStatus.PartitionLabel)
+		if baseOsStatus != nil {
+			baseOsSetPartitionInfoInStatus(ctxPtr, baseOsStatus, failedPartStatus.PartitionLabel)
+			publishBaseOsStatus(ctxPtr, baseOsStatus)
+		}
+		return
+	}
+	if ctxPtr.configUpdateRetry != retryUpdateCounter {
+		log.Noticef("handleUpdateRetryCounter: configUpdateRetry change: %d to %d",
+			ctxPtr.configUpdateRetry, retryUpdateCounter)
+		ctxPtr.configUpdateRetry = retryUpdateCounter
+		saveConfigRetryUpdateCounter(ctxPtr)
+	}
+	if ctxPtr.currentUpdateRetry != retryUpdateCounter {
+		log.Noticef("handleUpdateRetryCounter: currentUpdateRetry change: %d to %d",
+			ctxPtr.currentUpdateRetry, retryUpdateCounter)
+		ctxPtr.currentUpdateRetry = retryUpdateCounter
+		saveCurrentRetryUpdateCounter(ctxPtr)
+	}
+	publishBaseOSMgrStatus(ctxPtr)
+}
+
+func lookupBaseOsConfigByVersion(ctxPtr *baseOsMgrContext, shortVersion string) *types.BaseOsConfig {
+	sub := ctxPtr.subBaseOsConfig
+	items := sub.GetAll()
+	for _, cfg := range items {
+		baseOSConfig := cfg.(types.BaseOsConfig)
+		if baseOSConfig.BaseOsVersion == shortVersion {
+			return &baseOSConfig
+		}
+	}
+	return nil
+}
+
+func saveCurrentRetryUpdateCounter(ctxPtr *baseOsMgrContext) {
+	log.Functionf("saveCurrentRetryUpdateCounter (%s) - counter: %d",
+		currentRetryUpdateCounterFile, ctxPtr.currentUpdateRetry)
+	err := ioutil.WriteFile(currentRetryUpdateCounterFile,
+		[]byte(fmt.Sprintf("%d", ctxPtr.currentUpdateRetry)), 0644)
+	if err != nil {
+		log.Errorf("saveCurrentRetryUpdateCounter write to %s: %s", currentRetryUpdateCounterFile, err)
+	}
+}
+
+func saveConfigRetryUpdateCounter(ctxPtr *baseOsMgrContext) {
+	log.Functionf("saveConfigRetryUpdateCounter (%s) - counter: %d",
+		cfgRetryUpdateCounterFile, ctxPtr.configUpdateRetry)
+	err := ioutil.WriteFile(cfgRetryUpdateCounterFile,
+		[]byte(fmt.Sprintf("%d", ctxPtr.configUpdateRetry)), 0644)
+	if err != nil {
+		log.Errorf("saveConfigRetryUpdateCounter write to %s: %s", cfgRetryUpdateCounterFile, err)
+	}
+}
+
+func readSavedConfigRetryUpdateCounter() uint32 {
+	log.Trace("readSavedConfigRetryUpdateCounter - reading")
+	return readSavedRetryUpdateCounter(true)
+}
+
+func readSavedCurrentRetryUpdateCounter() uint32 {
+	log.Trace("readSavedCurrentRetryUpdateCounter - reading")
+	return readSavedRetryUpdateCounter(false)
+}
+
+// returns retryUpdateCounter if the file exists
+// else returns 0
+func readSavedRetryUpdateCounter(fromConfig bool) uint32 {
+	fileName := currentRetryUpdateCounterFile
+	if fromConfig {
+		fileName = cfgRetryUpdateCounterFile
+	}
+	log.Tracef("readSavedRetryUpdateCounter - reading %s", fileName)
+
+	b, err := ioutil.ReadFile(fileName)
```

Comment:
If we have some bug which causes these files in /persist to somehow grow to a GB, then you'll go in a OOM reboot loop forever.

I added a maxreadsize arg to fileutils.StatAndRead() to have an upper limit on how big buffers we allocate when we read files from /persist (since we did hit such an OOM loop for some other files in /persist)

Reply:
Thank you for details! Done.